src/server.rs

```diff
@@ -0,0 +1,374 @@
+//
+//   This Source Code Form is subject to the terms of the Mozilla Public
+//   License, v. 2.0. If a copy of the MPL was not distributed with this
+//   file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+
+use std::{collections::VecDeque, sync::Arc, time::Duration};
+
+use dashmap::DashMap;
+use mqtt_format::v3::{
+    connect_return::MConnectReturnCode, packet::MPacket, qos::MQualityOfService, strings::MString,
+    will::MLastWill,
+};
+use tokio::{
+    io::{AsyncWriteExt, DuplexStream, ReadHalf, WriteHalf},
+    net::{TcpListener, ToSocketAddrs},
+    sync::Mutex,
+};
+use tracing::{debug, error, info, trace};
+
+use crate::{
+    error::MqttError,
+    mqtt_stream::MqttStream,
+    topics::{ClientInformation, SubscriptionManager},
+    MqttPacket, PacketIOError,
+};
+
+#[derive(Debug, Clone)]
+pub struct MqttMessage {
+    author_id: Arc<ClientId>,
+    topic: String,
+    payload: Vec<u8>,
+    retain: bool,
+    qos: MQualityOfService,
+}
+
+impl MqttMessage {
+    fn from_last_will(will: &MLastWill, author_id: Arc<ClientId>) -> MqttMessage {
+        MqttMessage {
+            topic: will.topic.to_string(),
+            payload: will.payload.to_vec(),
+            qos: will.qos,
+            retain: will.retain,
+            author_id,
+        }
+    }
+
+    pub fn qos(&self) -> MQualityOfService {
+        self.qos
+    }
+
+    pub fn retain(&self) -> bool {
+        self.retain
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        self.payload.as_ref()
+    }
+
+    pub fn topic(&self) -> &str {
+        self.topic.as_ref()
+    }
+
+    pub fn author_id(&self) -> &ClientId {
+        self.author_id.as_ref()
+    }
+}
+
+pub struct MqttServer {
+    clients: DashMap<ClientId, ClientSession>,
+    client_source: ClientSource,
+    subscription_manager: SubscriptionManager,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ClientId(String);
+
+impl ClientId {
+    pub(crate) fn new(id: String) -> Self {
+        ClientId(id)
+    }
+}
+
+impl<'message> TryFrom<MString<'message>> for ClientId {
+    type Error = ClientError;
+
+    fn try_from(ms: MString<'message>) -> Result<Self, Self::Error> {
+        Ok(ClientId(ms.to_string()))
+    }
+}
```

Comment:
Does this actually have to be a `TryFrom`? Could it be a `From`?

Reply:
Right now yes, but in the future this might fail due to an invalid clientid